### ISSUE TITLE
ORBIT_DISPLAY

### ISSUE THREAD

**Issue by MOARdV** [MOARdV/AvionicsSystems#102]
Schematic rendering like the JSIOrbitDisplay, just using line renderers instead of inefficient GL calls.

Options:

Filled planet instead of a circle (make it easier to understand which line is orbit, which is planet.  Same for atmosphere.

**Comment by theonegalen** [comment]
Different colors for the filled planet depending on which planet we're orbiting?

**Comment by MOARdV** [comment]
> Different colors for the filled planet depending on which planet we're orbiting?

Yes.  And maybe a master override color if the IVA designer wants to control the color directly.  And, while I'm at it, maybe a cool fade effect for the atmosphere (although that's such a small slice, it may not be worth the extra work).

**Comment by Alexustas** [comment]
> Different colors for the filled planet depending on which planet we're orbiting?

It sounds interesting, but only if it does not require a lot of resources and we will be able to change or completely turn off the color fill. 

Personally, I use overlapping information text over an orbital map and color filling can make the text less readable, especially small text. so I do not think I'm going to use this feature extensively

**Comment by Alexustas** [comment]
What I personally would be interested in getting in the new version of the orbital map is a couple of features:

- A render of some mark of a given longitude, or more correctly, of its projection onto a map. To give the user at least a general idea of the location of the vessel relative to the surface. as the source of the longitude value, there may be:
  - zero meridian (light gray color)
  - longitude of the target (yellow line if the target is the ship landed on the surface, or a selected Waypoint)
  - longitude specified by the user (for example, longitude of KSC or any other random location, blue line),     
  - and also the longitude of the predicted landing site (if it exists)

![bandicam 2017-12-28 02-56-25-700](https://user-images.githubusercontent.com/6713671/34394420-c892721a-eb7a-11e7-95ec-b8a8fd0b1799.jpg)

- the second feature is the ability to switch the focus of the map to the next body, if the result of the current trajectory or planned maneuver is an encounter

![bandicam 2017-12-28 03-07-26-713](https://user-images.githubusercontent.com/6713671/34394627-456596ea-eb7c-11e7-9a1c-42c98d68f45a.jpg)

**Comment by nevercast** [comment]
Seeing upcoming orbits would be awesome for planning chained maneuvers. One thing I miss in my IVA is my excessive abuse of conics on the map.

**Comment by MOARdV** [comment]
> Regarding OP, it's a surprise to me that GL Calls are inefficient, why is the line renderer faster?

GL calls are done in immediate mode, which is an artifact of the pre-GPU OpenGL API.  Modern GPUs don't support immediate mode, so the GL API or the game engine above it must batch and send that data.  Those calls are done every Unity Update() call (including the transforms to put them into the right place on-screen), so that's a lot of recomputing the same data over and over at the mod's level, and sending that data through Unity.  The LineRenderer (presumably) is not emulating immediate mode, and it does not require resending data every Update.

**Comment by nevercast** [comment]
That's horrible! Thanks for the information, I would have kept acting on the assumption that using GL is direct hardware - wasn't aware Unity was using such a dated GL API

**Comment by MOARdV** [comment]
A lot of stuff to do for this task.

I need to decide how to handle target orbit rendering.  I think RPM ignored the argument of periapsis for the orbits it rendered, but I really should account for it - if the target is in an elliptical orbit with a periapsis on the opposite side of the planet from the ownship's periapsis, the relative orbits will be very wrong if I do not account for relative arguments of periapsis.  I also need to work out how to handle updating orbit lines - I don't want to update them every Update(), or even every FixedUpdate().  Maybe I can store the orbital parameters in a vector and test them on FixedUpdate() to see if they have changed.

**Comment by MOARdV** [comment]
Ah!  Here's something: since the orbits will be drawn with LineRenderer, I can allow each line to have two colors (start and end), similar to the ELLIPSE page object.  With two colors, I can draw the orbits similar to how the Map View draws them, with startColor where the vessel currently is, and endColor at the end of the orbit (future position).

**Comment by MOARdV** [comment]
Very early support for ORBIT_DISPLAY on [DropBox](https://www.dropbox.com/s/lgffr6rtdf0cc6b/AvionicsSystems-0.14.0b2.zip?dl=0).  Support for current vessel (for eccentricity < 1.0) and drawing the body that the vessel orbits.

**Comment by theonegalen** [comment]
My big request for this feature (and all displays) is the ability to go completely monochrome to emulate an early CRT monitor. Greenscale, that horrible Orangescale I remember from library computers in the 90s, Grayscale... The orbit display with RPM seems not to have the ability to change the color of the body itself, which means the greenscale CRT MFD I'm working on has white in it.

**Comment by MOARdV** [comment]
> My big request for this feature (and all displays) is the ability to go completely monochrome to emulate an early CRT monitor.

My intent is that you can configure all of the components in the page config - you can override the atmosphere and body colors, set the various orbit colors, etc.  That's actually available with the release I did today.  [See](https://github.com/MOARdV/AvionicsSystems/wiki/MASMonitor#orbit_display) the documentation for a little more detail.
